cpp/src/parquet/column_reader.h

```diff
@@ -278,9 +278,13 @@ class PARQUET_EXPORT RecordReader {
 
   /// \brief Attempt to read indicated number of records from column chunk
   /// Note that for repeated fields, a record may have more than one value
-  /// and all of them are read.
+  /// and all of them are read. If read_dense_for_nullable is true, it will
+  /// not leave any space for null values. Otherwise, it will read spaced.
+  /// Readers must call Reset() before switching between reading dense and
+  /// spaced since reading dense will not update the valid_bits_.
   /// \return number of records read
-  virtual int64_t ReadRecords(int64_t num_records) = 0;
+  virtual int64_t ReadRecords(int64_t num_records,
+                              bool read_dense_for_nullable = false) = 0;
```

Comment:
Thinking about this API, is there a reason you can think of for someone between to switch between Dense/Non-Dense?  Would it make sense pass this as an option to Make instead?

Reply:
Can we derive the null_count info from ColumnChunkMetaData or PageHeader and automatically enable it? To be conservative, we cannot enable it if null_count stats if missing. This would be much safer and user-friendly.